src/models/ListRule.ts

```diff
@@ -16,32 +16,235 @@ limitations under the License.
-     * Recommendations are normalised to their stable types.
+     * A glob for `entity`.
      */
-    public get recommendation(): string|null {
-        if (RECOMMENDATION_BAN_TYPES.includes(this.action)) return RECOMMENDATION_BAN;
-        return null;
+    private glob: MatrixGlob;
+    constructor(
+        /**
+         * The entity covered by this rule, e.g. a glob user ID, a room ID, a server domain.
+         */
+        public readonly entity: string,
+        /**
+         * A human-readable reason for this rule, for audit purposes.
+         */
+        public readonly reason: string,
+        /**
+         * The type of entity for this rule, e.g. user, server domain, etc.
+         */
+        public readonly kind: EntityType,
+        /**
+         * The recommendation for this rule, e.g. "ban" or "opinion", or `null`
+         * if the recommendation is one that Mjölnir doesn't understand.
+         */
+        public readonly recommendation: Recommendation | null) {
+        this.glob = new MatrixGlob(entity);
     }
 
+    /**
+     * Determine whether this rule should apply to a given entity.
+     */
     public isMatch(entity: string): boolean {
         return this.glob.test(entity);
     }
+
+    /**
+     * Validate and parse an event into a ListRule.
+     *
+     * @param event An *untrusted* event.
+     * @returns null if the ListRule is invalid or not recognized by Mjölnir.
+     */
+    public static parse(event: {type: string, content: any}): ListRule | null {
+        // Parse common fields.
+        // If a field is ill-formed, discard the rule.
+        const content = event['content'];
+        if (!content || typeof content !== "object") {
+            return null;
+        }
+        const entity = content['entity'];
+        if (!entity || typeof entity !== "string") {
+            return null;
+        }
+        const recommendation = content['recommendation'];
+        if (!recommendation || typeof recommendation !== "string") {
+            return null;
+        }
+
+        const reason = content['reason'] || '<no reason>';
+        if (typeof reason !== "string") {
+            return null;
+        }
+
+        let type = event['type'];
+        let kind;
+        if (USER_RULE_TYPES.includes(type)) {
+            kind = EntityType.RULE_USER;
+        } else if (ROOM_RULE_TYPES.includes(type)) {
+            kind = EntityType.RULE_ROOM;
+        } else if (SERVER_RULE_TYPES.includes(type)) {
+            kind = EntityType.RULE_SERVER;
+        } else {
+            return null;
+        }
+
+        // From this point, we may need specific fields
+        if (RECOMMENDATION_BAN_VARIANTS.includes(recommendation)) {
+            return new ListRuleBan(entity, reason, kind);
+        } else if (RECOMMENDATION_OPINION_VARIANTS.includes(recommendation)) {
+            let opinion = parseInt(content['opinion'], 10);
```

Comment:
Hold on a minute, why are we using `parseInt` here?

Reply:
@Gnuxie What else would we use?